x-pack/plugins/task_manager/server/routes/health.ts

```diff
@@ -0,0 +1,169 @@
+import { Logger, ServiceStatus, ServiceStatusLevels } from '../../../../../src/core/server';
+import {
+  MonitoringStats,
+  summarizeMonitoringStats,
+  HealthStatus,
+  RawMonitoringStats,
+} from '../monitoring';
+import { TaskManagerConfig } from '../config';
+
+type MonitoredHealth = RawMonitoringStats & { id: string; status: HealthStatus; timestamp: string };
+
+const LEVEL_SUMMARY = {
+  [ServiceStatusLevels.available.toString()]: 'Task Manager is healthy',
+  [ServiceStatusLevels.degraded.toString()]: 'Task Manager is unhealthy',
+  [ServiceStatusLevels.unavailable.toString()]: 'Task Manager is unavailable',
+};
+
+export function healthRoute(
+  router: IRouter,
+  monitoringStats$: Observable<MonitoringStats>,
+  logger: Logger,
+  taskManagerId: string,
+  config: TaskManagerConfig
+): Observable<ServiceStatus> {
+  // if "hot" health stats are any more stale than monitored_stats_required_freshness (pollInterval +1s buffer by default)
+  // consider the system unhealthy
+  const requiredHotStatsFreshness: number = config.monitored_stats_required_freshness;
+
+  // if "cold" health stats are any more stale than the configured refresh (+ a buffer), consider the system unhealthy
+  const requiredColdStatsFreshness: number = config.monitored_aggregated_stats_refresh_rate * 1.5;
+
+  function calculateStatus(monitoredStats: MonitoringStats): MonitoredHealth {
+    const now = Date.now();
+    const timestamp = new Date(now).toISOString();
+    const summarizedStats = summarizeMonitoringStats(monitoredStats, config);
+
+    /**
+     * If the monitored stats aren't fresh, return a red status
+     */
+    const healthStatus =
+      hasStatus(summarizedStats.stats, HealthStatus.Error) ||
+      hasExpiredHotTimestamps(summarizedStats, now, requiredHotStatsFreshness) ||
+      hasExpiredColdTimestamps(summarizedStats, now, requiredColdStatsFreshness)
+        ? HealthStatus.Error
+        : hasStatus(summarizedStats.stats, HealthStatus.Warning)
+        ? HealthStatus.Warning
+        : HealthStatus.OK;
+    return { id: taskManagerId, timestamp, status: healthStatus, ...summarizedStats };
+  }
+
+  const serviceStatus$: Subject<ServiceStatus> = new Subject<ServiceStatus>();
+
+  /* keep track of last health summary, as we'll return that to the next call to _health */
+  let lastMonitoredStats: MonitoringStats | null = null;
+
+  /* Log Task Manager stats as a Debug log line at a fixed interval */
+  monitoringStats$
+    .pipe(
+      throttleTime(requiredHotStatsFreshness),
+      tap((stats) => {
+        lastMonitoredStats = stats;
+      }),
+      // Only calculate the summerized stats (calculates all runnign averages and evaluates state)
+      // when needed by throttling down to the requiredHotStatsFreshness
+      map((stats) => withServiceStatus(calculateStatus(stats)))
+    )
+    .subscribe(([monitoredHealth, serviceStatus]) => {
+      serviceStatus$.next(serviceStatus);
+      logger.debug(JSON.stringify(monitoredHealth));
```

Comment:
**nit** I wonder if we can add something before the `JSON.stringify` to indicate what the data is.

Ex:
```suggestion
      logger.debug(`Latest task manager stats: ${JSON.stringify(monitoredHealth)}`);
```

Reply:
I actually left it out on purpose to make it easier to grep these out, but TBH that might be early optimisation seeing as we haven't got structured logs in Kibana at all yet.

Reply:
Added `Latest Monitored Stats: ` as TM is already mentioned, for example:

```
[2020-10-27T09:59:27.851Z][DEBUG][plugins.taskManager] Latest Monitored Stats: {"id":"5b2de169-2785-441b-ae8c-186a1936b17d","timestamp":"2020-10-27T09:59:27.837Z","status":"OK","last_update":"2020-10-27T09:59:27.837Z","stats":{"configuration":{"timestamp":"2020-10-27T09:59:25.273Z","value":{"request_capacity":1000,"max_poll_inactivity_cycles":10,"monitored_aggregated_stats_refresh_rate":60000,"monitored_stats_running_average_window":50,"monitored_task_execution_thresholds":{"default":{"error_threshold":90,"warn_threshold":80},"custom":{}},"poll_interval":1000,"max_workers":10},"status":"OK"},"runtime":{"timestamp":"2020-10-27T09:59:27.837Z","value":{"polling":{"last_successful_poll":"2020-10-27T09:59:27.837Z","result_frequency_percent_as_number":{"NoTasksClaimed":100,"RanOutOfCapacity":0,"PoolFilled":0}},"drift":{"p50":null,"p90":null,"p95":null,"p99":null},"execution":{"duration":{},"result_frequency_percent_as_number":{}}},"status":"OK"},"workload":{"timestamp":"2020-10-27T09:59:26.152Z","value":{"count":0,"task_types":{},"schedule":[],"overdue":0,"estimated_schedule_density":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]},"status":"OK"}}}
```